atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/queue/KvsSweepQueueScrubber.java

```diff
@@ -31,35 +31,47 @@ public KvsSweepQueueScrubber(SweepableCells cells, SweepableTimestamps timestamp
      * Cleans up all the sweep queue data from the last update of progress up to and including the given sweep
      * partition. Then, updates the sweep queue progress.
      * @param shardStrategy shard and strategy to scrub for.
-     * @param previousProgress previous last partition sweep has processed.
+     * @param oldProgress previous last partition sweep has processed.
      * @param newProgress last partition sweep has processed.
      */
-    public void scrub(ShardAndStrategy shardStrategy, long previousProgress, long newProgress) {
-        scrubSweepableCells(shardStrategy, newProgress);
-        scrubSweepableTimestamps(shardStrategy, previousProgress, newProgress);
+    public void scrub(ShardAndStrategy shardStrategy, long oldProgress, long newProgress) {
+        scrubSweepableCells(shardStrategy, oldProgress, newProgress);
+        scrubSweepableTimestamps(shardStrategy, oldProgress, newProgress);
         progressTo(shardStrategy, newProgress);
     }
 
-    private void scrubSweepableCells(ShardAndStrategy shardStrategy, long newProgress) {
-        scrubDedicatedRows(shardStrategy, newProgress);
-        scrubNonDedicatedRow(shardStrategy, newProgress);
+    private void scrubSweepableCells(ShardAndStrategy shardStrategy, long oldProgress, long newProgress) {
+        if (oldProgress < 0) {
+            return;
+        }
+        long oldLastSweptPartition = SweepQueueUtils.tsPartitionFine(oldProgress);
+        long newMinSweepPartition = SweepQueueUtils.tsPartitionFine(newProgress + 1);
+        if (newMinSweepPartition > oldLastSweptPartition) {
+            scrubDedicatedRows(shardStrategy, oldLastSweptPartition);
+            scrubNonDedicatedRow(shardStrategy, oldLastSweptPartition);
+        }
     }
 
-    private void scrubDedicatedRows(ShardAndStrategy shardStrategy, long partition) {
-        sweepableCells.deleteDedicatedRows(shardStrategy, partition);
+    private void scrubDedicatedRows(ShardAndStrategy shardStrategy, long partitionToDelete) {
+        sweepableCells.deleteDedicatedRows(shardStrategy, partitionToDelete);
     }
 
-    private void scrubNonDedicatedRow(ShardAndStrategy shardStrategy, long partition) {
-        sweepableCells.deleteNonDedicatedRow(shardStrategy, partition);
+    private void scrubNonDedicatedRow(ShardAndStrategy shardStrategy, long partitionToDelete) {
+        sweepableCells.deleteNonDedicatedRow(shardStrategy, partitionToDelete);
     }
 
-    private void scrubSweepableTimestamps(ShardAndStrategy shardStrategy, long oldPartition, long newPartition) {
-        if (SweepQueueUtils.partitionFineToCoarse(newPartition) > SweepQueueUtils.partitionFineToCoarse(oldPartition)) {
-            sweepableTimestamps.deleteRow(shardStrategy, oldPartition);
+    private void scrubSweepableTimestamps(ShardAndStrategy shardStrategy, long oldProgress, long newProgress) {
+        if (oldProgress < 0) {
+            return;
```

Comment:
This is an awkward edge case that arises from the initial previous swept timestamp being -1. It is actually not a correctness issue, but without this check as long as we reach at least the end of first partition in our first sweep we will send an unnecessary request to delete the first partition which is inconsistent with how we behave otherwise.

Reply:
Makes sense. The refactored version is crystal clear, thanks!